### DISCUSSION TITLE
Make search results more relevant by more precise keyword matching

### DISCUSSION THREAD

**Discussion by ralphrmartin** [searxng/searxng#2529] [Ideas]
**Is your feature request related to a problem? Please describe.**
Many searches return too many irrelevant results. It would be useful to be able to exclude result pages containing one or more irrelevant keywords. For example, this can be done in Google searches using "-keyword". Other search engines may also provide such an option, but perhaps using different syntax e.g. "NOT keyword".

**Describe the solution you'd like**
Searxng could define and use some syntax for this purpose in its searches, internally converting them into the appropriate syntax for each search engine consulted.

**Describe alternatives you've considered**
An alternative would be to filter results returned by the search engines to exclude ones containing unwanted keywords, but that would put much more load on searxng.

**Additional context**
A related feature, which would also be extremely useful, would be to only include results which **explicitly** contain searched for keywords (not near matches). This used to be possible with Google using "+keyword", but alas no longer, so this probably would need postfiltering in searxng.

**Comment by intrepidsilence** [comment]
Is this being considered for addition? This is a very important feature to make this engine usable.

**Reply by return42** [reply]
Its not planned / SearXNG is a meta-search engine, it does not have an index.

> For example, this can be done in Google searches using "-keyword". 

If you explicitly select the google engine, then you can pass such parameters to it.  For example, search the word "app" without "store":

    !go app -store

**Comment by ralphrmartin** [comment]
I do not wish to restrict searches to Google in this way. I'd like such an option passed to all search engines which support such a feature.

**Reply by return42** [reply]
SearXNG is an aggregator and not all of the resources it aggregates does have this option.

**Comment by ralphrmartin** [comment]
It would still be helpful to cut out some of the unwanted results from services which do have the option.

**Reply by return42** [reply]
See my comment above https://github.com/searxng/searxng/discussions/2529#discussioncomment-6311862

**Comment by ralphrmartin** [comment]
I don't really understand what you mean by your reply?.  Searxng could pass such an option to to the services (search engines) which support such an option (like Google), which would mean that the results being aggregated contain fewer unwanted results from those sources which support such an option, making the aggregated results more useful. 
Searxng could use a syntax of its own to denote unwanted words, and then edit the queries sent to the services to use the appropriate syntax for each services which does support such an option, or quietly forget the unwanted word for ones which do not support it.

**Reply by return42** [reply]
> or quietly forget the unwanted word for ones which do not support it.

but the results from those support this option and those engines which doesn't will be mixed in one result list .. in the aggregation you will have *unwanted results* .. so where is the benefit?

**Comment by ralphrmartin** [comment]
The benefit is that there will be _fewer_ results with the unwanted words, assuming that the search engines do not return  identical results to each other - if they do, there is no point in aggregating results anyway. One could always restrict the search to engines which do support such an option.

**Reply by return42** [reply]
> One could always restrict the search to engines which do support such an option.

Theoretically this would be feasible, but in practice it raises a lot of questions ... e.g. on the user side: which syntax would be the most suitable, Google's? 

--> https://quickref.me/google-search.html

Assuming we would implement the search syntax of Google, then the users would also expect other options like `daterange:`,  `site:`, `AND/OR` and many more .. To generalize this syntax for an aggregator would not be an easy task, since there are syntactical features in one search index (Google), but in the other possibly not (Startpage?) ... In addition, the instance operators themselves decide which search services are active in the instance ... some turn off e.g. google completely / so what

- if the users uses a syntactic feature that is only supported by google, but google-engine is inactive?

- what means `+` (Force exact match on single phrase) in the different search services / when we aggregate?

These are just two exemplary questions, if you think about it for a while, you will come up with at least 20-30 more questions and some of them will not be answered meaningfully by an aggregator, at least not in all possible setups and constellations.

In summary, the interests of the users, the interests of the instance operators and the syntactic features of the search engines must be combined in one application / in one syntax.

I don't see a way to reconcile all this and get adequate benefit from it, given the massive investment -- which would be required.

This is why such an feature is not planned .. but we are open to proposals that are less conflictual, feasible in practice, and easily accessible to the user.

In the meantime a search syntax like `!go app -store` (I suggested above) is a good workaround.

**Comment by ralphrmartin** [comment]
I'm sure users would be happy to go with whatever straightforward syntax you suggest that could not be mistaken for part of the query. While further operators like daterange:, site: would be useful, they are perhaps less important than excluding unwanted junk.  One way to approach this could be to have a table, which says for each search engine and each operator whether the engine supports the operator, and how to turn the user's query including an operator into to a query that works for that search engine (using a rewriting rule based on regular expressions for example). This could be used for each operator in turn.

 I note that restricted  time ranges are already supported somewhat, by a pop up to restrict the results shown to Anytime, Last Day, Week, Month, Year..

The user could have a choice whether to ignore search engines that do not support a given operator, or whether to ignore the operator for such search engines.

Aggregating results all of which contain an exact phrase is not seemingly a problem. - just remove duplicates as you add the results tp a list of results, which is presumably done anyway.

Even if dealing with all operators is too tricky or too much effort. It would still be useful to just have a way of expressing "exclude results with _xxx_" in them.  This would be a **massive** benefit for most users, I suspect, particularly in cases where a word has been "hijacked" by use as the name of a product or company, or is taken to be a misspelling of a more common word, or is replaced by a presumed synonym.
